### PR TITLE
new Measurement and Metric objects tickets/DM-9542

### PR DESCRIPTION
**Pull request by parejkoj**
Update Measurement and Metric objects to match design in sqr-017, and add output_measurements method for Tasks to use.

### CODE REVIEW

**Comment by jonathansick** [comment]
@parejkoj Should I be worried about any of these test failures?

```
tests/test_job.py

Traceback (most recent call last):
  File "tests/test_job.py", line 10, in <module>
    from lsst.validate.base import (MeasurementBase, Metric, Datum, BlobBase,
ImportError: cannot import name MeasurementBase
tests/test_measurements.py

....E...
======================================================================
ERROR: test_create_invalid_unit (__main__.MeasurementTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_measurements.py", line 79, in test_create_invalid_unit
    with self.assertRaises(u.UnitTypeError):
AttributeError: 'module' object has no attribute 'UnitTypeError'

----------------------------------------------------------------------
Ran 8 tests in 0.001s

FAILED (errors=1)
tests/test_output_measurements.py

E
======================================================================
ERROR: test_output_measurements_example (__main__.OutputTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_output_measurements.py", line 32, in test_output_measurements_example
    self.outfile = lsst.validate.base.output_measurements(self._testMethodName, data)
  File "/Users/jsick/lsst/qa-system/validate_base/python/lsst/validate/base/output.py", line 12, in output_measurements
    measurements.append(Measurement('.'.join((package_name, name)), value))
  File "/Users/jsick/lsst/qa-system/validate_base/python/lsst/validate/base/measurement.py", line 110, in __init__
    raise NotImplementedError('Cannot autoload validate_metrics yet')
NotImplementedError: Cannot autoload validate_metrics yet

----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
The following tests failed:
/Users/jsick/lsst/qa-system/validate_base/tests/.tests/test_job.py.failed
/Users/jsick/lsst/qa-system/validate_base/tests/.tests/test_measurements.py.failed
/Users/jsick/lsst/qa-system/validate_base/tests/.tests/test_output_measurements.py.failed
3 tests failed
scons: *** [checkTestStatus] Error 1
scons: building terminated because of errors.
```

File: python/lsst/validate/base/metric.py [hunk 1]
```diff
@@ -1,21 +1,112 @@
...
 
-__all__ = ['Metric', 'load_metrics']
+
+class MetricRepo(object):
+    """A collection of MetricSets, each identified by name.
+
+    Parameters
+    ----------
+    path : `str`
+        path to the directory defining this MetricRepo
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Make this a sentence (capital and full stop).

File: python/lsst/validate/base/metric.py [hunk 2]
```diff
@@ -1,21 +1,112 @@
...
+    metric_sets : `dict`
+        dict of name: `MetricSet`
+    """
+    path = None
+    """`str` giving the path to the directory defining this MetricRepo."""
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Not a standard yet, but I tend to write these as

```
Path of the directory defining this MetricRepo (`str`).
```

File: python/lsst/validate/base/metric.py [hunk 3]
```diff
@@ -1,21 +1,112 @@
...
+
+class MetricRepo(object):
+    """A collection of MetricSets, each identified by name.
+
+    Parameters
+    ----------
+    path : `str`
+        path to the directory defining this MetricRepo
+    metric_sets : `dict`
+        dict of name: `MetricSet`
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]:
```
Dictionary of ``name: MetricSet`` key-value pairs.
```

File: python/lsst/validate/base/metric.py [hunk 4]
```diff
@@ -1,21 +1,112 @@
...
+
+    metric_sets = None
+    """`dict` of all the MetricSets defined in this repo, identified by name."""
+
+    def __init__(self, path, metric_sets):
+        self.path = path
+        self.metric_sets = metric_sets
+
+    @classmethod
+    def from_path(cls, path):
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: What do you think of `from_metric_dir`?

File: python/lsst/validate/base/metric.py [hunk 5]
```diff
@@ -1,21 +1,112 @@
...
+        The name of this metric set, usually the name of an LSST package
+    metrics : `dict` of `str`: `Metric`
+        A `dict` of names and their associated metrics.
+    """
+
+    name = None
+    """Name of the MetricSet, usually the name of a package."""
+
+    metrics = None
+    """{name: metric} dict of the metrics"""
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Better to put the literals in double backticks,

```
``{name : metric}``
```

and "metric" ->

```
`Metric` instances
```

File: python/lsst/validate/base/metric.py [hunk 6]
```diff
@@ -1,21 +1,112 @@
...
+
+class MetricSet(object):
+    """A collection of metrics, each identified by name.
+
+    Parameters
+    ----------
+    name : `str`
+        The name of this metric set, usually the name of an LSST package
+    metrics : `dict` of `str`: `Metric`
+        A `dict` of names and their associated metrics.
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]:
```
`Metric`\ s.
```

File: python/lsst/validate/base/metric.py [hunk 7]
```diff
@@ -135,55 +205,18 @@ def from_yaml(cls, metric_name, yaml_doc=None, yaml_path=None,
...
-                    deps[name] = d
-            spec = Specification(name=spec_doc['level'],
-                                 quantity=spec_doc['value'],
-                                 unit=spec_doc['unit'],
-                                 filter_names=spec_doc.get('filter_names', None),
-                                 dependencies=deps)
-            m.specs.append(spec)
-
+        # NOTE: description is a folded block, which gets an appended '\n'.
+        # NOTE: a fix for this would be to denote the block with '>-' in the .yaml.
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py:212]: That's fair. We can change the existing samples in `validate_metrics` to use the chomp modifier, but I think there's no harm always stripping trailing newlines.

**Recommedation**: `strip` -> `rstrip`.

File: python/lsst/validate/base/metric.py [hunk 8]
```diff
@@ -200,26 +233,36 @@ def from_json(cls, json_data):
...
+                json_data['unit'],
                 reference_doc=json_data['reference']['doc'],
                 reference_page=json_data['reference']['page'],
                 reference_url=json_data['reference']['url'])
         return m
 
-    def __getattr__(self, key):
-        if key in self.parameters:
-            return self.parameters[key]
+    # def __getattr__(self, key):
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Might as well delete this now.

File: python/lsst/validate/base/output.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+# See COPYRIGHT file at the top of the source tree.
+from __future__ import absolute_import, division, print_function
+
+import json
+
+from .measurement import MeasurementSet, Measurement
+
+
+def output_measurements(package_name, measurement_dict):
```
Comment (jonathansick) [inline python/lsst/validate/base/output.py:9]: I'm not so sure about this API. I think it's more likely we'll put this in `Job` instead. We can leave this in for now, but I'm fairly certain it'll get deprecated before the epic is out.

File: bin.src/validate_metrics_repo.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+import lsst.pex.exceptions
+import lsst.utils
+from lsst.validate.base import MetricRepo
+
+
+def main():
+    try:
+        default_metrics_dir = os.path.join(lsst.utils.getPackageDir('validate_metrics'), 'metrics')
+    except lsst.pex.exceptions.NotFoundError:
```
Comment (jonathansick) [inline bin.src/validate_metrics_repo.py]: Honest question, what does `lsst.pex.exceptions` do for us and the user here? why can't we use an `OSError` or a custom subclass of `OSError`? From my perspective outside Science Pipelines, at least, it doesn't seem useful to add a dependency for a C++ exception library into a pure Python project.

**Comment by parejkoj** [comment]
Did you see my note on the review request? I pretty much only worked on the Measurement/Metric portion, leaving refactoring the other bits to someone else (e.g., once we sort out the API for a Job).

File: bin.src/validate_metrics_repo.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
+"""
+Validate the contents of a RetricsRepo, by constructing all of its
```
Comment (jonathansick) [inline bin.src/validate_metrics_repo.py]: You could rename this to `lint_metrics_repo.py`, especially if we add static schema checking with jsonschema to help users identify errors more precisely.

File: bin.src/validate_metrics_repo.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+                        help="directory containing the MetricRepo to be checked.")
+    args = parser.parse_args()
+
+    metric_repo = MetricRepo.from_path(args.repo)
+
+    print("Successfully constructed a MetricRepo from {}".format(args.repo))
+    print()
+    print(metric_repo)
+
+if __name__ == "__main__":
```
Comment (jonathansick) [inline bin.src/validate_metrics_repo.py]: Needs an extra space above (E305).

File: python/lsst/validate/base/measurement.py [hunk 13]
```diff
@@ -1,46 +1,74 @@
...
 
-class MeasurementBase(with_metaclass(abc.ABCMeta,
-                                     type('NewBase',
-                                          (QuantityAttributeMixin,
-                                           JsonSerializationMixin,
-                                           DatumAttributeMixin),
-                                          {}))):
-    """Base class for Measurement classes.
+class MeasurementSet(object):
+    """A collection of Measurements of Metrics, associated with a MetricSet
```
Comment (jonathansick) [inline python/lsst/validate/base/measurement.py]: Add a full stop.

File: python/lsst/validate/base/measurement.py [hunk 14]
```diff
@@ -1,46 +1,74 @@
...
+        The name of this `MetricSet` (usually the name of a package).
+    measurements : `dict` of `str`: `astropy.Quantity`
+        The measurements (astropy.Quantities) of each metric, keyed on the
+        metric name, to be looked up in metric_set.
+    job : `Job`, optional
+        The Job that produced these `Measurements`, linking them to their
+        provenance and other metadata.
+    metric_set : MetricSet, optional
+        A `MetricSet` to extract the metric definitions from. If None, use
+        validate_metrics[name].
```
Comment (jonathansick) [inline python/lsst/validate/base/measurement.py]: You mean that `validate_metrics` is used by default? Just be aware that you're mixing metaphors here, `validate_metrics` is a Git repository, but not a Python object that's subscriptable. we can fix this docstring later, though.

**Comment by jonathansick** [comment]
Ah, I missed that comment about the failing tests. Okay, the rest of my comments are coming right up.

File: python/lsst/validate/base/measurement.py [hunk 15]
```diff
@@ -1,46 +1,74 @@
...
+    job = None
+    """`Job` that this MeasurementSet was produced by."""
 
-    Subclasses must (at least) implement the following attributes:
+    def __init__(self, name, measurements, job=None, metric_set=None):
+        if metric_set is None:
+            raise NotImplementedError('Cannot autoload validate_metrics yet')
+        self.name = name
+        self.measurements = {}
+        self.job = job
```
Comment (jonathansick) [inline python/lsst/validate/base/measurement.py:47]: Do you intend to do anything with `Job` yet? Like, does the `MeasurementSet` assign itself as a `measurement_set` attribute of the `Job`?

File: python/lsst/validate/base/metric.py [hunk 16]
```diff
@@ -1,21 +1,112 @@
...
+    def from_yaml(cls, name, yaml):
+        metrics = {}
+        for subname in yaml:
+            metric_name = "{}.{}".format(name, subname)
+            m = Metric.from_yaml(subname, yaml_doc=yaml)
+            metrics[metric_name] = m
+        return cls(name, metrics)
+
+    @classmethod
+    def from_list(cls, name, metric_list):
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: What's the user of the `from_list` API? I couldn't find it being used yet in the code base.

File: python/lsst/validate/base/metric.py [hunk 17]
```diff
@@ -64,39 +157,16 @@ class Metric(JsonSerializationMixin):
...
-    """
-
-    def __init__(self, name, description, operator_str,
-                 specs=None, parameters=None,
+    def __init__(self, name, description, unit, tags=None,
                  reference_doc=None, reference_url=None, reference_page=None):
         self.name = name
         self.description = description
+        self.unit = unit
+        self.tags = tags
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Might want it do

```python
if tags is None:
    self.tags = []
else:
    self.tags = tags
```

but maybe I'm missing a use-case for tags being None rather than length-zero.

File: python/lsst/validate/base/metric.py [hunk 18]
```diff
@@ -200,26 +233,36 @@ def from_json(cls, json_data):
...
-        if key in self.parameters:
-            return self.parameters[key]
+    # def __getattr__(self, key):
+    #     if key in self.parameters:
+    #         return self.parameters[key]
+    #     else:
+    #         raise AttributeError("%r object has no attribute %r" %
+    #                              (self.__class__, key))
+
+    def validate(self, quantity):
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py]: Maybe rename to `check_unit`?

Comment (jonathansick) [reply python/lsst/validate/base/metric.py]: Specifications will have a `check` method that tests whether a measurement has passed a specification or so, so there would be nice symmetry with that.

File: python/lsst/validate/base/metric.py [hunk 19]
```diff
@@ -55,6 +142,12 @@ class Metric(JsonSerializationMixin):
     description = None
     """Short description of the metric (`str`)."""
 
+    unit = None
```
Comment (jonathansick) [inline python/lsst/validate/base/metric.py:145]: Heads up that in `Specification`-land I'm distinguishing between the string and Astropy Unit object forms.

- `unit` is an `astropy.units.Unit` instance
- `unit_str` is the string representation of a `Unit`. Maybe you could do this here too?

**Review by jonathansick — COMMENTED** [review]
I think I'm fine with merging this. I've got various comments that you can address as you please. I've also got some suggestions for renaming things.

Probably the biggest thing is that I'm not entirely sure of is the role of the `output` module and how it relates to the `Job` object. But I don't think there's harm in merging it as is and just moving things forward.

File: python/lsst/validate/base/metric.py (hunk 16 above)
Comment (parejkoj) [reply python/lsst/validate/base/metric.py]: I use it in test_measurements, to build a MetricSet from a list of metrics.

File: python/lsst/validate/base/metric.py (hunk 19 above)
Comment (parejkoj) [reply python/lsst/validate/base/metric.py:145]: Ah, great! I was wondering about how exactly to manage this, so I'm glad you already came up with something. I'll see if I can switch it over to that.

File: python/lsst/validate/base/metric.py (hunk 17 above)
Comment (parejkoj) [reply python/lsst/validate/base/metric.py]: Good idea. Unless we know of a use-case now, lets just enforce `list`-ness.

File: python/lsst/validate/base/metric.py (hunk 8 above)
Comment (parejkoj) [reply python/lsst/validate/base/metric.py]: Wasn't sure if  you thought you might have use for it. It's gone now.

File: python/lsst/validate/base/output.py (hunk 9 above)
Comment (parejkoj) [reply python/lsst/validate/base/output.py:9]: Yeah. Oddly, this was the original topic for this ticket, but I had to do the metric and measurement stuff to get it to make sense, at which point it faded from relevancy.

We'll probably want to rethink the whole output system, possibly to use the butler metadata system or something, so it plugs into other bits of Tasks trivially.

File: bin.src/validate_metrics_repo.py (hunk 10 above)
Comment (parejkoj) [reply bin.src/validate_metrics_repo.py]: It's not what it gains us, but it's the fact that `lsst.utils.getPackageDir` raises `pex.NotFoundError`. There are plans to make the pex exceptions look like standard python exceptions (see, e.g. DM-9435), but we're stuck with it for now if we want to use `getPackageDir`, and I think we do.

Alternately, file a ticket against `lsst.utils` about how `getPackageDir` should raise a more standard python type and I'll back you on it.

File: python/lsst/validate/base/measurement.py (hunk 14 above)
Comment (parejkoj) [reply python/lsst/validate/base/measurement.py]: I clarified my intention in the docstring.

File: python/lsst/validate/base/measurement.py (hunk 15 above)
Comment (parejkoj) [reply python/lsst/validate/base/measurement.py:47]: No. We never fully fleshed out what a Job should contain (we added MeasurementSet late in our planning), so I'm not sure how this all should behave.

**Comment by parejkoj** [comment]
I've pushed an update that I think takes care of your suggestions.

File: python/lsst/validate/base/metric.py (hunk 16 above)
Comment (jonathansick) [reply python/lsst/validate/base/metric.py]: What if the `metrics` argument in  the `__init__` is a list? Then in the `__init__` itself we make `self.metrics` the dict, using the metric's `name` as the key?

That has the nice advantage of:

1. users aren't duplicating `metric.name` and the key in their input to `MetricSet`.
2. Don't need the `from_list` class method.

Note, I'm building a `Name` class in a separate PR that I'll expedite that will make it easy to get a fully-qualified metric name since I recognize that you don't want `metric.name` itself.

File: python/lsst/validate/base/measurement.py [hunk 20]
```diff
@@ -1,46 +1,74 @@
...
+    """`str` the name of the `MetricSet` these `Measurement`s are of."""
+
+    measurements = None
+    """`dict` of all `Measurement` names to `Measurement`s."""
+
+    job = None
+    """`Job` that this MeasurementSet was produced by."""
 
-    Subclasses must (at least) implement the following attributes:
+    def __init__(self, name, measurements, job=None, metric_set=None):
```
Comment (jonathansick) [inline python/lsst/validate/base/measurement.py:42]: Like in [my comment](https://github.com/lsst/validate_base/pull/5/files#r106524171) about `MetricSet`, I'm wondering if the `measurements` argument here could just be a list, and then in `__init__` we get the fully-qualified metric name from the measurement objects to build `self.measurements` as a list.

File: python/lsst/validate/base/measurement.py [hunk 21]
```diff
@@ -302,19 +328,23 @@ def check_spec(self, name):
...
 
+    def __eq__(self, other):
+        return (self.value == other.value) and (self.metric == other.metric)
 
-class DeserializedMeasurement(MeasurementBase):
+
+class DeserializedMeasurement(Measurement):
     """Measurement deserialized from JSON.
 
+    TODO: do we still need this one?
```
Comment (jonathansick) [inline python/lsst/validate/base/measurement.py]: Good question. Probably not. My goal was for it to be possible to rebuild a Measurement from its serialized JSON output. The `DeserializedMeasurement` was a compromise because I realized I couldn't rebuild the exact same Blob class (with compute methods and such) from its JSON alone.

tl;dr It's safe to delete this and we'll figure out how deserialize `validate_drp` classes from their JSON serializations.

File: python/lsst/validate/base/output.py (hunk 9 above)
Comment (jonathansick) [reply python/lsst/validate/base/output.py:9]: I'm losing my mind, sorry. You're right that `output_measurements` as a shim to the Butler is probably a good thing. But keeper in mind that there's more to a measurement than the value (need to connect to blobs, and to Job data).

Previously I used `Job.json` to recursively serialize all Measurement, Blob, and Job data via their `json` attributes. So I think we'll want to think about how to bring that idea back.

(I understand that right now MeasurementSet is effectively being serialized via it's `__str__` method rather than a `json` attribute calling `Measurement` `json` attributes.)

I'm fine if we merge this now and then re-group with more concrete design thinking on persistence.

**Comment by jonathansick** [comment]
To summarize my new comments:

1. I'm wondering if `MetricSet` and `MeasurementSet` can take lists rather than dicts, and then build the internal dictionaries themselves. See https://github.com/lsst/validate_base/pull/5#discussion_r106524171

2. The `output_measurements` function forgoes the recursive `json` attribute machinery we had before (that's my reading of it). We can merge this now as a proof-of-concept but we'll want to come back and think about an appropriate way to serialize measurements, their blobs, and Job context information. SQUASH will need all of this info. see https://github.com/lsst/validate_base/pull/5#discussion_r106534874

File: python/lsst/validate/base/metric.py (hunk 16 above)
Comment (parejkoj) [reply python/lsst/validate/base/metric.py]: I've changed it so `__init__` takes either a list or a dict: I think it needs a dict so that `from_yaml` can pass it a fully-constructed thing. If your new `Name` takes care of all that "automatically", that'd be great.

File: python/lsst/validate/base/output.py (hunk 9 above)
Comment (parejkoj) [reply python/lsst/validate/base/output.py:9]: Ok, cool. Especially since we're merging to an in-progress epic branch, we can sort out details as other parts are finished.

File: python/lsst/validate/base/measurement.py (hunk 20 above)
Comment (parejkoj) [reply python/lsst/validate/base/measurement.py:42]: It made more sense for Metrics, I think: a Measurement is (currently) just a name and a value, so getting a dict of those works well. You can feel free to redo it once you have a `Name` helper in place that handles the full name conversion.
